### PR TITLE
Export from Google3

### PR DESCRIPTION
**Pull request by alexames**

### CODE REVIEW

**Comment by jonsimantov** [comment]
My understanding is that we would not do another large export from g3 to here, but instead we would simply look at the relevant commits that went in since the last merge and apply the changes by hand.

**Review by jonsimantov — COMMENTED** [review]
@cynthiajoan Could you look over each file's changes under remote_config to ensure that it captures the correct state of the source code?
@dconeybe Could you look over each file's changes under firestore to ensure that this PR captures the correct state of the source code?
@alexames I added some ? comments where the merge seemed like it was doing something odd, could you double check that those are needed and aren't just undoing something that was done in GitHub on purpose? I think there are a number of things that we purposely fixed on GitHub that are being undone by this merge (and I might have missed some).

It also doesn't fill me with confidence that the tests are broken. Also also, please run the two CPP Packaging and Integration Tests workflows on this PR's branch.

File: remote_config/src/common.cc [hunk 1]
```diff
@@ -23,57 +23,16 @@
 // Register the module initializer.
 FIREBASE_APP_REGISTER_CALLBACKS(remote_config,
                                 {
-                                  if (app == ::firebase::App::GetInstance()) {
```
[the remark this answers is not in the document]
Comment (alexames) [reply remote_config/src/common.cc:26]: I believe this is correct. The changes here are from February, which is much newer than the code that's being removed.

File: remote_config/src/common.h [hunk 2]
```diff
@@ -69,11 +71,32 @@ namespace internal {
 // Implemented in each platform module.
 bool IsInitialized();
 
-// Registers a cleanup task for this module if auto-initialization is disabled.
-void RegisterTerminateOnDefaultAppDestroy();
```
[the remark this answers is not in the document]
Comment (alexames) [reply remote_config/src/common.h:73]: I believe this is also correct, it was removed recently. @a-maurice, can you confirm? It looks like you were the one who removed this from google3

File: remote_config/src/common.h [hunk 3]
```diff
@@ -15,6 +15,8 @@
 #ifndef FIREBASE_REMOTE_CONFIG_CLIENT_CPP_SRC_COMMON_H_
 #define FIREBASE_REMOTE_CONFIG_CLIENT_CPP_SRC_COMMON_H_
 
+#include <type_traits>
```
[the remark this answers is not in the document]
Comment (alexames) [reply remote_config/src/common.h:18]: I can remove it if you'd like. It was added in google3 fairly recently, though it's not clear to me why.

**Comment by dconeybe** [comment]
@var-const Can you take a look at this PR also? All of your google3 changes to the C++ SDK submitted so far should be included (the cutoff CL was cl/360516349 and your last CL was cl/360506827).

**Comment by dconeybe** [comment]
Looks good for Firestore

**Review by dconeybe — APPROVED** [review]
LGTM for Firestore

**Review by var-const — APPROVED** [review]
LGTM. One question, though -- it seems like many of the commits being imported were already applied to `main`?

**Pull request closed without merging**